tls/s2n_ktls.c

```diff
@@ -86,9 +87,95 @@ S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_kt
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->client);
+    RESULT_ENSURE_REF(conn->server);
+    RESULT_ENSURE_REF(key_material);
+    RESULT_ENSURE_REF(crypto_info);
+    RESULT_ENSURE_REF(tls_tx_rx_mode);
+    RESULT_GUARD(s2n_ktls_validate(conn, ktls_mode));
+
+    /* TODO once other ciphers and protocols are supported, check the
+     * negotiated cipher (AES_128_GCM). This would involve adding a unique
+     * identifier to s2n_cipher.
+     */
+
+    /* set tls mode */
+    if (ktls_mode == S2N_KTLS_MODE_SEND) { /* TX */
+        *tls_tx_rx_mode = S2N_TLS_TX;
+    } else { /* RX */
+        *tls_tx_rx_mode = S2N_TLS_RX;
+    }
+
+    /* set values based on mode of operation */
+    uint8_t *implicit_iv = NULL;
+    uint8_t *sequence_number = NULL;
+    struct s2n_blob *key = NULL;
+
+    if ((conn->mode == S2N_SERVER && ktls_mode == S2N_KTLS_MODE_SEND)        /* server sending */
+            || (conn->mode == S2N_CLIENT && ktls_mode == S2N_KTLS_MODE_RECV) /* client receiving */
+    ) {
+        /* if server is sending or client is receiving then use server key material */
+        key = &key_material->server_key;
+        implicit_iv = conn->server->server_implicit_iv;
+        sequence_number = conn->server->server_sequence_number;
+    } else {
+        key = &key_material->client_key;
+        implicit_iv = conn->client->client_implicit_iv;
+        sequence_number = conn->client->client_sequence_number;
+    }
+
+    RESULT_GUARD(s2n_blob_validate(key));
+    RESULT_ENSURE_EQ(key->size, S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE);
+
+    /* set values on crypto_info */
+    crypto_info->info.cipher_type = S2N_TLS_CIPHER_AES_GCM_128;
+    crypto_info->info.version = S2N_TLS_1_2_VERSION;
+    RESULT_CHECKED_MEMCPY(crypto_info->salt, implicit_iv, S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->rec_seq, sequence_number, S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->key, key->data, S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->iv, implicit_iv, S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_ktls_set_keys(struct s2n_connection *conn, s2n_ktls_mode ktls_mode,
+        struct s2n_key_material *key_material)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(key_material);
+    RESULT_GUARD(s2n_ktls_validate(conn, ktls_mode));
+
+    int fd;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    /* Only AES_128_GCM for TLS 1.2 is supported at the moment. */
+    struct s2n_tls12_crypto_info_aes_gcm_128 crypto_info = { 0 };
+    int tls_tx_rx_mode = 0;
+    RESULT_GUARD(s2n_ktls_init_aes128_gcm_crypto_info(conn, ktls_mode, key_material, &crypto_info, &tls_tx_rx_mode));
+
+    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
+    if (s2n_in_unit_test()) {
+        return S2N_RESULT_OK;
+    }
```

Comment:
You've got this check in a couple places now. It might be cleaner / easier if you wrap setsockopt in a method that includes this check-- maybe even one that can call a test version of setsockopt for unit tests?

Reply:
There are only going to be 2 calls to setsockopt so I really dont think its worth having a wrapper around this and paying the cost of the indirection.